Initial content caching policy (#1174)
3scale/APIcast: gateway/src/apicast/policy/content_caching/Readme.md

# APICast Policy caching

> This policy allows to enable and disable caching based on customized conditions. These

Reviewer: From a UX perspective, should we clarify that there are 2 ENVs to configure this policy and their default values?

Author: That is defined in the #upstream response codes.

conditions can only be applied on the client request, where upstream responses
cannot be used in this policy.

> 

Reviewer: Should we clarify somewhere that this policy respects the cache headers set by upstream? By that, I mean that even if this policy is enabled and there's a rule that matches, a response will not be cached if the "Cache-control" header is set to "no-cache" by the upstream.

# APICast Policy caching

> This policy allows to enable/disable caching based on a custom conditions. These

Reviewer 2: ```suggestion
This policy allows to enable and disable caching based on customized conditions. These
```

conditions can only be applied on the client request, where upstream responses
cannot be used in this policy.

## Example configuration

```
{
  "name": "apicast.policy.content_caching",
  "version": "builtin",
  "configuration": {
    "rules": [
      {
        "cache": true,
        "header": "X-Cache-Status-POLICY",
        "condition": {
          "combine_op": "and",
          "operations": [
            {
              "left": "{{method}}",
              "left_type": "liquid",
              "op": "==",
              "right": "GET"
            }
          ]
        }
      }
    ]
  }
}
```

## Recommended configuration

> - For POST/PUT/DELETE method, the recommendation is not have this enabled.

Reviewer 2: ```suggestion
- Set `APICast Policy caching` as disabled, for any of the following methods: POST, PUT, DELETE.
```

- If one condition matches, and it enables the cache, the execution will be

>   stopped and it'll be not disabled. Sort by priority is important here.

Reviewer 2: ```suggestion
  stopped and it will be not disabled. Sort by priority is important here.
```

## Upstream response headers

> At the moment, nginx [`proxy_cache_valid`

Reviewer 2: ```suggestion
At the moment, the NGINX [`proxy_cache_valid`
```

directive](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_valid)
information can only be set globally, with the `APICAST_CACHE_STATUS_CODES` and

> `APICAST_CACHE_MAX_TIME`. If your upstream need to have a different behavior

Reviewer 2: ```suggestion
`APICAST_CACHE_MAX_TIME`. If your upstream requires a different behavior
```
